### DISCUSSION TITLE
Arbitrary CBOR tags

### DISCUSSION THREAD

**Discussion by Pique7** [danielaparker/jsoncons#567] [Q&A]
How can I specify an arbitrary (semantic) tag for a CBOR array (or any other CBOR data type), if at all possible?

**Comment by danielaparker** [comment]
Arbitrary CBOR tags are currently only supported for byte strings.

**Reply by Pique7** [reply]
OK, in my (simple) case I can also solve it by inserting the tag value when the encoding is done.
I've tried to understand how the CBOR encoder works. It doesn't seem too difficult to provide a public function for inserting tag values, maybe disregarding error handling ...

**Reply by Pique7** [reply]
My workarounds (for simple data) so far ...

**Encoding** example:
```
vector<uint8_t> data;
cbor::cbor_bytes_encoder encoder(data);
encoder.begin_array(1);
encoder.flush();
data.push_back(0xC1);   // tag value
encoder.uint64_value(1);  // actual data value
encoder.end_array();
encoder.flush();
```

**Decoding** (simplified) example:
```
cbor::cbor_bytes_cursor cursor(data);
cursor.next();
if (*(data.data() + cursor.column() - 2) == 0xC1) {
  // expected tag
  uint64_t myDataValue = cursor.current().get<uint64_t>();
}
```

I don't know whether there is a better way to deal with it, for me this seems to be sufficient at the moment.

**Reply by danielaparker** [reply]
Code review could be interesting :-)

I think we can add support for the encoding part in the next release, by adding non-virtual CBOR specific functions to `basic_cbor_encoding`, substituting the CBOR tag for the cross-library `jsoncons::semantic_tag`.

The decoding part requires more thought.

**Reply by danielaparker** [reply]
Writing and reading CBOR tags is now supported on master. Your example will look like

```cpp
#include <cassert>
#include <iostream>
#include <jsoncons_ext/cbor/cbor.hpp>

namespace cbor = jsoncons::cbor;

int main()
{
    std::vector<uint8_t> data;
    cbor::cbor_bytes_encoder encoder(data);
    encoder.begin_array(1);
    encoder.uint64_value_with_tag(1, 0xC1);
    encoder.end_array();
    encoder.flush();

    cbor::cbor_bytes_cursor cursor(data);
    assert(cursor.current().event_type() == jsoncons::staj_event_type::begin_array);
    cursor.next();
    assert(cursor.raw_tag() == 0xC1);
    assert(cursor.current().get<uint64_t>() == 1);
}
```
